text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,527 @@
+    ///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    ///         write!(f, "Example Error")
+    ///     }
+    /// }
+    ///
+    /// impl fakecore::error::Error for Error {
+    ///     fn provide_context<'a>(&'a self, mut request: Pin<&mut Request<'a>>) {
+    ///         request.provide::<Backtrace>(&self.backtrace);
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///     let backtrace = Backtrace::new();
+    ///     let error = Error { backtrace };
+    ///     let dyn_error = &error as &dyn fakecore::error::Error;
+    ///     let backtrace_ref = dyn_error.context::<Backtrace>().unwrap();
+    ///
+    ///     assert!(core::ptr::eq(&error.backtrace, backtrace_ref));
+    /// }
+    /// ```
+    fn provide_context<'a>(&'a self, request: Pin<&mut Request<'a>>) {}
+}
+```
+
+Example implementation:
+
+```rust
+fn provide_context<'a>(&'a self, mut request: Pin<&mut Request<'a>>) {
+    request
+        .provide::<Backtrace>(&self.backtrace)
+        .provide::<SpanTrace>(&self.span_trace)
+        .provide::<dyn Error>(&self.source)
+        .provide::<Vec<&'static Location<'static>>>(&self.locations)
+        .provide::<[&'static Location<'static>]>(&self.locations);
+}
+```
+
+Example usage:
+
+```rust
+let e: &dyn Error = &concrete_error;
+
+if let Some(bt) = e.context::<Backtrace>() {
+    println!("{}", bt);
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+In Rust, errors typically gather two forms of context when they are created:
+context for the *current error message* and context for the *final* *error
+report*. The `Error` trait exists to provide an interface to context intended
+for error reports. This context includes the error message, the source error,
+and, more recently, backtraces.
+
+However, the current approach of promoting each form of context to a method on
+the `Error` trait doesn't leave room for forms of context that are not commonly
+used, or forms of context that are defined outside of the standard library.
+
+## Extracting non-std types from `dyn Errors`
+
+By adding a generic form of these member access functions we are no longer
+restricted to types defined in the standard library. This opens the door to
+many new forms of error reporting.
+
+### Example use cases this enables
+
+* using alternatives to `std::backtrace::Backtrace` such as
+  `backtrace::Backtrace` or [`SpanTrace`]
+* zig-like Error Return Traces by extracting `Location` types from errors
+  gathered via `#[track_caller]` or similar.
+* error source trees instead of chains by accessing the source of an error as a
+  slice of errors rather than as a single error, such as a set of errors caused
+  when parsing a file TODO reword
+* Help text such as suggestions or warnings attached to an error report
+
+## Moving `Error` into `libcore`
+
+Adding a generic member access function to the `Error` trait and removing the
+`backtrace` function would make it possible to move the `Error` trait to libcore
```

Comment:
I had forgotten that `Error::backtrace` was still unstable -- might be worth making that explicit.